### PR TITLE
RPC: Transaction deser can be quite slow

### PR DESCRIPTION
**Pull request by t-nelson**
#### Problem

RPC accepts large (sometimes too) base58-encoded TXs.  Base58 is slow.

#### Summary of Changes

- Check encoded TX size before decoding
- Support submitting base64-encoded TXs (default to base58 for compat)

cc/ @jstarry for web3js support

### CODE REVIEW

**Comment by jstarry** [comment]
Awesome! We can switch web3 over by default once all clusters are running this change

File: client/src/rpc_config.rs [hunk 1]
```diff
@@ -11,12 +11,20 @@ pub struct RpcSignatureStatusConfig {
     pub search_transaction_history: bool,
 }
 
+#[derive(Clone, Copy, Debug, Deserialize, PartialEq, Serialize)]
+#[serde(rename_all = "camelCase")]
+pub enum RpcTransactionEncoding {
```
Comment (mvines) [inline client/src/rpc_config.rs:16]: I think I'd rather reuse `UiTransactionEncoding` here, and then just reject the additional encodings in code

File: client/src/rpc_client.rs [hunk 2]
```diff
@@ -48,6 +49,17 @@ pub struct RpcClient {
...
 }
 
+fn serialize_encode_transaction(
+    transaction: &Transaction,
+    encoding: RpcTransactionEncoding,
+) -> String {
+    let serialized = serialize(transaction).unwrap();
+    match encoding {
+        RpcTransactionEncoding::Base58 => bs58::encode(serialized).into_string(),
+        RpcTransactionEncoding::Base64 => base64::encode(serialized),
```
Comment (mvines) [inline client/src/rpc_client.rs:59]: Happy to only support base64 actually, and never let a Rust client send slomo base58

**Review by CriesofCarrots — COMMENTED** [review]
Nice! Docs?

File: client/src/rpc_client.rs (hunk 2 above)
Comment (CriesofCarrots) [reply client/src/rpc_client.rs:59]: +1

File: client/src/rpc_config.rs (hunk 1 above)
Comment (t-nelson) [reply client/src/rpc_config.rs:16]: Ah!  There _is_ a `UiTransactionEncoding`. I was looking in all of the wrong places :sweat_smile:

File: client/src/rpc_client.rs (hunk 2 above)
Comment (t-nelson) [reply client/src/rpc_client.rs:59]: Yep!  I think we need to roll that out over a couple releases to maintain our initiative to "break less stuff".  I'm planning to:
1. Add support for base64 here, maintaining base58 defaults and announcing base58 deprecation.
1. Drop base58 client-side in the next release, reiterating deprecation
1. Drop base58 server-side in the release after that, by now everyone's had fair warning.

I'll toss up some issues to track if you agree

Comment (mvines) [reply client/src/rpc_client.rs:59]: 👍

**Comment by t-nelson** [comment]
> Nice! Docs?

Here they are! https://github.com/solana-labs/solana/pull/12683/commits/584d682b62f5f3efa7015147854fb6ee74c894c2#diff-2301d6850e0d62de7061568b7dcaff47

File: client/src/rpc_config.rs (hunk 1 above)
Comment (t-nelson) [reply client/src/rpc_config.rs]: Updated in https://github.com/solana-labs/solana/pull/12683/commits/584d682b62f5f3efa7015147854fb6ee74c894c2

File: client/src/rpc_client.rs (hunk 2 above)
Comment (t-nelson) [reply client/src/rpc_client.rs]: Migration tracking issues at #12700 and #12701

**Review by mvines — APPROVED** [review]

**Review by CriesofCarrots — APPROVED** [review]

**Pull request merged**